### PR TITLE
for loops are now generated inline rather than in an IIFE

### PR DESCRIPTION
**Pull request by alubbe**
This again improves performance a little bit (part of https://github.com/jadejs/jade/issues/1975).

The first change is to rename ```$index``` to ```$$iN```, where N is the Nth for loop. The PR assumes this is safe. It also assumes it is safe to use variables that look like ```KEY__obj``` and ```KEY__l```, where KEY is either ```$$iN``` or the KEY provided by the user.

### CODE REVIEW

File: lib/lexer.js [hunk 1]
```diff
@@ -561,7 +562,8 @@ Lexer.prototype = {
     if (captures = /^(?:- *)?(?:each|for) +([a-zA-Z_$][\w$]*)(?: *, *([a-zA-Z_$][\w$]*))? * in *([^\n]+)/.exec(this.input)) {
       this.consume(captures[0].length);
       var tok = this.tok('each', captures[1]);
-      tok.key = captures[2] || '$index';
+      tok.key = captures[2] || '$$i' + this.eachCount;
+      this.eachCount++;
```
Comment (alubbe) [inline lib/lexer.js]: this is split into two lines on purpose for readability

**Comment by ForbesLindesay** [comment]
This is great, but definitely a breaking change.  Could you make this pull request against [jade-code-gen](https://github.com/jadejs/jade-code-gen) project as well.

**Comment by alubbe** [comment]
Moved variable name generation out of the lexer.
As a side effect, all variable names now start with ```jade_```, so if we consider those safe, the PR is now backwards-compatible.

**Comment by ForbesLindesay** [comment]
Unfortunately I still don't consider this backwards compatible because:

1. The immediately invoked function expression changes the scope of any variables the user defines within it.
2. The `$index` variable that was previously implicitly created, could be used by users.  I'm happy to break these users as part of the 2.0.0 release, but I don't want to break them as a point release.

I'm very excited for getting this done in the 2.0.0 release though :)

**Comment by alubbe** [comment]
Now that we have started landing breaking changes in master, can we merge this or should it go straight into jade-code-gen?

**Comment by ForbesLindesay** [comment]
It needs to go straight into jade-code-gen

**Pull request closed without merging**